### ISSUE TITLE
CSS Tailwind slow build time even with different files...

### ISSUE THREAD

**Issue by frederikhors** [sveltejs/svelte-preprocess#275]
I'm using Tailwind with default Svelte template and svelte-preprocess.

When I modify a single css the build is very slow (~10 seconds).

I read this: https://nystudio107.com/blog/speeding-up-tailwind-css-builds but I can't make it work with svelte-preprocess.

Also with different .css files imported in one `App.svelte` file when modifying just one of it (the lighter one, not the one with tailwindcss/utilities in it) the build is of the same time.

Why?

Thanks for your precious work.

**Comment by acoyfellow** [comment]
Tailwind 2.0 has crushed my development flow. Startup time has 50x'd, and my dev server is running out of memory and crashing in new ways I've yet to see.

**Comment by kaisermann** [comment]
Hey @frederikhors 👋 Could you provide a minimal repro? Would help a lot 😁

**Comment by natecorkish** [comment]
> Hey @frederikhors 👋 Could you provide a minimal repro? Would help a lot 😁

https://github.com/natecorkish/frontend-svelte/

Checkout to `7b7f13e7dea17e6421c38fc7e8b51a02197e2493` which has an inline tailwind stylesheet. You should get the same issue as the original poster.

**Comment by dionysiusmarquis** [comment]
I tested some scenarios always copy pasted the **generated css** coming from `@tailwind base; …` into a `<style global>…`:
Default sapper template with rollup: "vanilla" **~2s** with prebuilt tailwind css **~26s**
Typescript sapper template with rollup: "vanilla" **~4s** with prebuilt tailwind css **~39s**

When also using postcss and not a prebuilt css I'll have "~1m"

I'd say one of the main problems is that svelte already processes the whole tailwind css as it where "handwritten". An option to skip the whole precessing part for a specific `<script>` inside a `.svelte` file could help e.g.

**Comment by dionysiusmarquis** [comment]
After I investigated this deeper in the context of sapper. I found out that it's not wrong to consider it as bad practice to import styles inside `<style>` rather than `<script>` (following: https://github.com/sveltejs/svelte/issues/5745).
My conclusion is that you should most likely never add heavy "third party" css to svelte's `<style>` content. It's preferable to import them inside `<script>` and add the corresponding rollup plugins or webpack loaders to handle `.css`, `.pcss` imports.
This way you have better control of which styles are actually svelte component's css that should be processed by svelte and which styles are (in terms of Tailwind even pretty heavy) "third party" styles, that aren't bound to any specific svelte component. You just have to remember to also add a css minifier and enable tailwind puge to prod postcss

:heavy_exclamation_mark: (even though it's done this way in quite a lot blog posts regarding Tailwind in svelte)
```jsx
<styles lang="postcss">
  @import 'src/styles/tailwind.pcss'
…
```

:white_check_mark:
```jsx
<script>
  import 'src/styles/tailwind.pcss'
…
```

**Comment by frederikhors** [comment]
@dionysiusmarquis does this solve the problem?

**Comment by dionysiusmarquis** [comment]
Only tested it in the context of sapper and I'm down to **~5s** per `tailwind.pcss` change after moving the imports to script. With style imports it was at **>1m**. 
So for me it was a huge build time decrease. And that makes total sense, as svelte and svelte-preprocess don't need to "mess" with the Tailwind styles at all anymore.

**Comment by frederikhors** [comment]
@dionysiusmarquis what rollup's plugin are you using for `.pcss`?

**Comment by frederikhors** [comment]
@dionysiusmarquis can you share a gist for us, please?

**Comment by non25** [comment]
This has nothing to do with svelte-preprocess.
Here's the wasted computation pipeline with default svelte-preprocess & tailwind postcss config:

1. svelte-preprocess launches postcss with its default config on any style tag, which in this case makes tailwind AST for `@apply` and friends, which are not going to be used in every component
2. resulting css goes to the svelte compiler
3. svelte-loader or rollup-plugin-svelte stores all component styles from the compiler in memory cache
4. every compiled svelte component loads this style from cache through `import` statements
5. for webpack, due to how copypasted webpack-virtual-modules work, svelte-loader with `emitCss: true` reloads that virtual css file for every component on any change applying postcss transformations from default postcss config for example, making tailwind AST for the second time

So, make separate tailwind config for postcss working only for `lang="tailwind"` explicitly, that way you will be only making tailwind's AST where it's needed.

```javascript
const tailwindcss = require('tailwindcss')('./tailwind.config.js');
...
preprocess: sveltePreprocess({
  postcss({ content, filename, attributes }) {
    // if lang is tailwind, then skip postcss default config processing for
    // this style tag and just return it as is (postcss is launched for any
    // style tag by default in svelte-preprocess to make :global work)
    return attributes.lang === 'tailwind' ?
      { code: content } :
      postcss(preprocessPostCSSPlugins)
        .process(content, { from: filename })
        .then(({ css: code, map }) => ({ code, map }));
  },
  tailwind({ content, filename }) {
    // apply tailwind's transformation and then other postcss transformations
    // like autoprefixer for lang="tailwind" in one batch
    // this is more efficient than parsing style tags twice
    return postcss([ tailwindcss, ...preprocessPostCSSPlugins ])
      .process(content, { from: filename })
      .then(({ css: code, map }) => ({ code, map }));
  }
})
```

If you are using same global postcss config with tailwind and autoprefixer for svelte-preprocess and `*.css` files, then you are making tailwind's AST twice, one for svelte-preprocess, and second is for loading compiled css from compiled svelte component making it slow as hell.

So I would advise you to make separate configs for tailwind and other postcss stuff like autoprefixer for svelte components and external files (something like `*.tcss`).

I don't know how to fix it in rollup, but this fixes reloading every virtual css file with global postcss config if postcss is configured for `*.css` files for webpack 4: https://github.com/sveltejs/svelte-loader/pull/133 making this playable.

Similar effect can be achieved using emitCss: false in dev mode: `emitCss: prod`.

I believe the smoothest DX with hmr (and this is important for style tags) can be achieved with webpack 5 for now.
Here's patch for `svelte-loader-hot` making use of webpack-virtual-modules as external dependency and loading it correct way as a plugin.

```diff
diff --git a/index.js b/index.js
index bfbdf583eac6fd9b0ebc9c187be34d531f95980c..8d7e8a2907dbf20a959f97a5b447c2d494cbc1b8 100644
--- a/index.js
+++ b/index.js
@@ -1,6 +1,6 @@
 const { basename, extname, relative } = require('path');
 const { getOptions } = require('loader-utils');
-const VirtualModules = require('./lib/virtual');
+const VirtualModules = require('webpack-virtual-modules');
 const posixify = require('./lib/posixify');
 
 const {
@@ -68,15 +68,9 @@ function deprecatePreprocessOptions(options) {
 	options.preprocess = options.preprocess || preprocessOptions;
 }
 
-const virtualModuleInstances = new Map();
-
-module.exports = function(source, map) {
-	if (this._compiler && !virtualModuleInstances.has(this._compiler)) {
-		virtualModuleInstances.set(this._compiler, new VirtualModules(this._compiler));
-	}
-
-	const virtualModules = virtualModuleInstances.get(this._compiler);
+let virtualModules;
 
+function loader(source, map) {
 	this.cacheable();
 
 	const options = Object.assign({}, getOptions(this));
@@ -154,3 +148,14 @@ module.exports = function(source, map) {
 		callback(new Error(`${err.name}: ${err.toString()}`));
 	});
 };
+
+class SveltePlugin extends VirtualModules {
+	constructor(modules) {
+		super(modules);
+		virtualModules = this;
+	}
+}
+
+loader.plugin = SveltePlugin;
+
+module.exports = loader;
diff --git a/package.json b/package.json
index 14dd944538313ebcd8cd12736afe17c04d457e31..df901f66b897d9afb7439a1b99d60510dbedde6b 100644
--- a/package.json
+++ b/package.json
@@ -27,7 +27,8 @@
 	"dependencies": {
 		"loader-utils": "^1.1.0",
 		"svelte-dev-helper": "^1.1.9",
-		"svelte-hmr": "^0.10.3"
+		"svelte-hmr": "^0.10.3",
+		"webpack-virtual-modules": "^0.4.1"
 	},
 	"devDependencies": {
 		"chai": "^4.1.2",
```

Changes needed to webpack config:

```javascript
const SveltePlugin = require('svelte-loader-hot').plugin;
...
{
  test: /\.svelte$/,
  use: {
    loader: 'svelte-loader-hot',
    options: {
      dev: !prod,
      emitCss: prod,
      hotReload: !prod,
      hotOptions: {
        noPreserveState: true,
        optimistic: false
      }
    },
    ...
  }
},
...
{
  // required to prevent errors from Svelte on Webpack 5+
  test: /svelte\/.*\.mjs$/,
  resolve: {
    fullySpecified: false
  }
}
...
plugins: [
  new MiniCssExtractPlugin({
    filename: '[name].css'
  }),
  new SveltePlugin()
],
```

Let me know if you are interested in svelte webpack 5 hmr boilerplate with bearable tailwind configuration.

**Comment by non25** [comment]
@frederikhors what do you think about the template above?

**Comment by benmccann** [comment]
@non25 thanks for all the info - and of course your help on `svelte-loader`!  I'm afraid I didn't quite follow the explanation. I have little understanding of many of these pieces having done very little with Svelte CSS, but if you don't mind bearing with me I'll do my best to help since I know Tailwind is popular and this is a common pain point. I'm also more familiar with the `rollup-plugin-svelte` than `svelte-loader`, so maybe we can start there just so I can wrap my head around this all

Here's my understand of everything and a bunch questions. Please correct me where I'm wrong and don't assume I know much about Svelte CSS :smile: 

Tailwind introduces new directives like `@apply` that are used in css files. Tailwind offers a postcss plugin to handle these directives. It also offers components which can be `@import`ed. To use Tailwind in Svelte components you must setup postcss with tailwind's plugin as a svelte preprocessor. Svelte then process the css to add scoping.

Is the main issue with Tailwind being slow that it has a large component library or that the Tailwind Postcss plugin is slow to start or run?

Most Svelte components won't import any Tailwind components. Is this issue primarily surfaced in the main svelte component for the app (e.g. `_layout.svelte` in Sapper)?

If we offer a `lang="tailwind"` this is allowing templates with vanilla CSS or postcss without tailwind to be faster, but it does not speed up templates that use Tailwind since they still need to do all the processing. I'm assuming most users who adopt Tailwind will use it in most of their templates, so how would this really help?

If making Tailwind AST is slow and we cut it down from doing it twice to doing it once we'd cut the time in half, but that might still be slow?

> So I would advise you to make separate configs for tailwind and other postcss stuff like autoprefixer for svelte components and external files (something like *.tcss).

You used "and" here a few times and am not quite sure how you're recommending to split it up. It should be split into only two preprocessor configs? What goes in each? E.g. #1 "tailwind and other postcss stuff like autoprefixer". #2 "external files". Or #1 "tailwind". #2 "other postcss stuff like autoprefixer" and "external files"

Should Tailwind stuff go in `global.css` or is it better to have it in the main svelte component?

What changes would you recommend to `rollup-plugin-svelte`?

**Comment by benmccann** [comment]
I've been reading through the issues in the Tailwind repo. This definitely isn't a problem unique to Svelte. A few things I found there that I'm wondering if people have tried:
* Upgrade to Tailwind 2.0.2+ and [split up your imports as described in this article](https://nystudio107.com/blog/speeding-up-tailwind-css-builds)
* Disable CSS source maps
* Try enabling/disabling `purge` of `scss` files (or all files?)

**Comment by non25** [comment]
@benmccann The article you found describes stuff better. :grin: 

My above post's main focus is currently irrelevant.
I've done tests and came to such conclusions because I was using one guy's config, which had:

- bug from `svelte-loader`, reloading every component's virtual css on any project change (irrelevant today)
- `.css` rule running postcss with tailwind, making reloading bug noticeable
- huge imports of tailwind base stuff
- purgecss & cssnano working in both modes


### Here's main takeaways I see now

#### Include huge css from tailwind into your bundle only in production

These ones:

```postcss
@tailwind base;
@tailwind components;
@tailwind utilities;
```

Prebuild them to be loaded in `index.html` for devmode and import & purge them in production.

In webpack, I [achieve it](https://github.com/non25/svelte-tailwind-template/blob/e5d7ed268b9a243e66d5955975b051e890393d14/webpack.config.js#L74-L79) with the help of `null-loader`, which outputs nothing from imports, and does so conditionally.

#### Launch postcss with tailwind only on css where you will write `@apply`s, etc.

Avoid making rules for `.css` that launch postcss with tailwind. With `emitCss: true` it will spin it's `@apply`s for preprocess, and then for imported virtual css.

This can go unnoticed because most guides recommend making one `postcss.config.js` for every postcss task and put tailwind plugin there.

I'd argue that people using tailwind would also use third-party svelte components and would want to run autoprefixer on them too, but without tailwind's overhead.

That's when explicit `lang=tailwind`, or tailwind files (`.tcss` for example) come in handy.

#### Sidenotes

Would be cool if there was an option for postcss/webpack, that saved spinned up stuff for `@apply`s in memory after getting done with one component and reusing it on the next. Cold build times would skyrocket.

Or if we could somehow collect all css from every component, preprocess it in one batch with tailwind and then do regular processing without launching tailwind plugin anymore.

**Comment by dionysiusmarquis** [comment]
I did a quite deep dive into the svelte css part. This preprocessor will add `global(…)` to every tailwind generated css definitions. Svelte then has an internal process to determine unused css parts etc. This processes can get quite heavy when executed on bigger generated, non hand written css like Tailwind. This processes are 100% unnecessary computations for generated css like Tailwind. As already stated, the only vanilla way atm to bypass this unnecessary processing is to import css in `script` and not `style`.

You can find the breakdown and example REPL here (it shifted to sapper but is still relevant in this context): https://github.com/sveltejs/svelte/issues/5745
tldr: Svelte css preprocessing/processing treads Tailwind css as "hand written" component css. As war as I can tell the most feasible solution on svelte level could be to add a way to tell svelte that the heavy, non hand written css parts shouldn't be processed by svelte at all

**Comment by non25** [comment]
@dionysiusmarquis I don't believe that is the case unless you use something like `@tailwind utilities` or `@import tailwindcss/utilities.css` in svelte component's style tag, which is completely unnecessary.

Also I don't think importing anything that is not mixins, vars, functions in svelte component's style tag is a viable choice.

Tailwind works in svelte components by replacing directives with css code. There shouldn't be any extraneous classes whatsoever.

**Comment by dionysiusmarquis** [comment]
> I don't believe that is the case unless you use something like @tailwind utilities or @import tailwindcss/utilities.css in svelte component's style tag, which is completely unnecessary.

But that's exactly the way everyone seem to use css libraries Tailwind in svelte intuitively. That is most likely also the reason why nearly every blog post will tell you to use tailwind in svelte this way. And to match this intuition a way to write or add `plain` css could be a logical next step, imo

**Comment by non25** [comment]
> That is most likely also the reason why nearly every blog post will tell you to use tailwind in svelte this way.

This is outrageous.

So for every component where you want to slap some tailwind's classes such articles will instruct you to import the whole thing ?
And in the end you will have like **~component_count** dups of whole tailwind's ~3mb csses ? Cool stuff.

Can you give links to such articles ?

**Comment by dionysiusmarquis** [comment]
In svelte terms you are pretty much always in "component territory", i'd say. The current way of dealing with general css definitions seems to be adding a `<style global …` inside `App.svelte` or `Layout.svelte` ("components"). Which is understandable and will generate only one global definition for all components, but svelte will kick in all css processing routines nevertheless. There is currently no way to tell svelte that incoming css definitions inside a `<style>` don't need any additional processing and should just added to the global definitions unchanged.
The only way would be to do it via the `<scripts>` pipeline with corresponding loader or plugin. But this will also create an additional css file you have to hash and apply to your main template as well.

**Comment by dionysiusmarquis** [comment]
> @dionysiusmarquis can you share a gist for us, please?

@frederikhors Here is a simple example. It's not really the way you should organise the css files when working with Tailwind, but it outlines my point. You should get expected "fast" build times and you should also be able to use all the Tailwind features as expected.

Here are the changes applied to the current plain `svelte-template`:
https://github.com/dionysiusmarquis/svelte-tailwind-template/commit/e053f8de8a33c8a2425025479b2da7ec19566adb

**Comment by kaisermann** [comment]
As @dionysiusmarquis described, the preprocessor always treats your CSS considering a component context, so having the preprocessor putting those `:global` and svelte parsing and ignoring all of Tailwind's declarations will definitely take a toll on one's machine. I'd suggest importing the base of your Tailwind CSS through JS or some other mean.
